scripts/backout_data_collection.py

```diff
@@ -0,0 +1,226 @@
+import json
+import logging
+import os
+from collections.abc import Generator
+from datetime import datetime, timedelta
+
+from tqdm import tqdm
+
+from bugbug import bugzilla, db, repository
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def download_databases() -> None:
+    logger.info("Cloning Mercurial database...")
+    repository.clone(repo_dir="hg_dir")
+
+    logger.info("Downloading bugs database...")
+    assert db.download(bugzilla.BUGS_DB)
+
+    logger.info("Downloading commits database...")
+    assert db.download(repository.COMMITS_DB, support_files_too=True)
+
+
+def preprocess_commits_and_bugs() -> tuple[dict, dict, dict]:
+    logger.info("Preprocessing commits and bugs...")
+    commit_dict, bug_dict = {}, {}
+    bug_to_commit_dict: dict[int, list] = {}
+
+    for commit in repository.get_commits(
+        include_no_bug=True, include_backouts=True, include_ignored=True
+    ):
+        commit_data = {
+            key: commit[key]
+            for key in ["node", "bug_id", "pushdate", "backedoutby", "backsout"]
+        }
+        commit_dict[commit["node"]] = commit_data
+
+        bug_to_commit_dict.setdefault(commit["bug_id"], []).append(commit_data)
+
+    # We only require the bug's resolution (to check if it is 'FIXED').
+    bug_dict = {
+        bug["id"]: bug["resolution"] for bug in bugzilla.get_bugs(include_invalid=True)
+    }
+
+    return commit_dict, bug_to_commit_dict, bug_dict
+
+
+def has_conflicts(diff: str) -> bool:
+    """Return True if the diff contains any conflict markers. Used with merge-tool ':fail'."""
+    conflict_markers = ["<<<<<<<", "=======", ">>>>>>>"]
+    return any(marker in diff for marker in conflict_markers)
+
+
+def generate_datapoints(
+    commit_limit: int,
+    commit_dict: dict,
+    bug_to_commit_dict: dict,
+    bug_dict: dict,
+    repo_dir: str,
+) -> Generator[dict, None, None]:
+    counter = 0
+    commit_limit = min(commit_limit, 709458)
+
+    logger.info("Generating datapoints...")
+
+    for commit in tqdm(
+        repository.get_commits(
+            include_no_bug=True, include_backouts=True, include_ignored=True
+        )
+    ):
+        counter += 1
+
+        bug_info = bug_dict.get(commit["bug_id"])
```

Comment:
The names here are a bit confusing. The `bug_inff` variable has only the resolution, so it could be more suitable to call it something like `bug_resolution`. Also, `bug_dict` is very general; I would use something more specific like `bug_resolution_map`.

Reply:
Done in 37c51b640aa4b5f38ab88e1fa1e4b4bd8fc525c3!